### ISSUE TITLE
Feature request: Fallback for `-sASYNCIFY=2`

### ISSUE THREAD

**Issue by hoodmane** [emscripten-core/emscripten#18652]
I'm very excited about using JavaScript promise integration once it starts gaining runtime support. I notice that it is an all or nothing thing right now: if we compile with `-sASYNCIFY=2` then the emitted code won't run at all on a runtime missing JSPI. It would be great to feature detect JSPI and have a fallback in case it is missing. For example this would be very useful for code splitting: If JSPI is present, we can lazily load code with `EM_JS_ASYNC`, if it is missing we can fall back to eagerly loading everything at initialization time.

If JSPI is missing, `instrumentWasmImports` could make calling the imports into an error, and `instrumentWasmExports` could call the `export$func` wrapper function with a dummy value for the `externref` first argument and wrap the result in `Promise.resolve`.

If we also want to work in the case that `externref` support is missing, we would have to dynamically generate the import and export wrappers. This should be similar to `createDyncallWrapper` but simpler.

I am happy to implement this if you are interested in supporting it.

**Comment by kripken** [comment]
Interesting idea!

I think it's possible. The runtime would need to check for JSPI and not call `Asyncify.instrumentWasmImports` and `Exports`, I assume, but hopefully not much else.

cc @brendandahl @tlively for thoughts.

**Comment by tlively** [comment]
Wouldn't turning import calls into errors make it very difficult to make this fallback graceful at the application level? It seems like a more general fallback would be to decide whether to download the normal JSPI or Asyncify-instrumented module depending on whether JSPI is avilable.

**Comment by hoodmane** [comment]
For my use cases, the costs of `-sASYNCIFY=1` are too high. I also do not want to have two entirely different wasm modules since that means that most tests have to be run in both configurations in order to ensure that neither has dynamic linking bugs. On the other hand, I am perfectly okay with saying that certain APIs fail unless the runtime supports JSPI. In the code splitting example, the `EM_JS_ASYNC` imports never get called when JSPI is missing because we eagerly loaded everything.

**Comment by brendandahl** [comment]
I don't think it makes sense to have this as a default behavior since an application could end up in a semi-broken state. I guess an option to opt into this could be fine though.

**Comment by hoodmane** [comment]
Yeah it is probably better in many cases to bail on startup rather than to start doing something that definitely won't work.
